### PR TITLE
libbpf-tools: add bindsnoop

### PR DESCRIPTION
**Pull request by chenhengqi**
```shell
sudo ./bindsnoop -t   
TIME(s)  PID     COMM             RET PROTO OPTS  IF   PORT  ADDR                                            
19:08:19 3944603 nc               0   TCP   ...Rr 0    1216  0.0.0.0                                         
19:08:43 3944709 nginx            0   TCP   ...R. 0    80    0.0.0.0                                         
19:08:43 3944709 nginx            0   TCP   ...R. 0    80    ::
```

Signed-off-by: Hengqi Chen <chenhengqi@outlook.com>

### CODE REVIEW

File: libbpf-tools/bindsnoop.bpf.c [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	event.pid = pid;
+	event.port = sport;
+	event.bound_dev_if = BPF_CORE_READ(sock, __sk_common.skc_bound_dev_if);
+	event.ret = ret;
+	event.proto = BPF_CORE_READ_BITFIELD_PROBED(sock, sk_protocol);
+	bpf_get_current_comm(&event.task, sizeof(event.task));
+	if (ver == 4) {
+		event.ver = ver;
+		bpf_probe_read_kernel(&event.addr, sizeof(event.addr), &inet_sock->inet_saddr);
+	} else {
```
Comment (davemarchevsky) [inline libbpf-tools/bindsnoop.bpf.c:97]: Mind adding a `// ver == 6` comment here for readability?

**Review by davemarchevsky — CHANGES_REQUESTED** [review]
I'm guessing strict compatibility with the `bcc-py` bindsnoop's output format is not a goal? e.g. The example output you provided differs from same invocation of `bindsnoop.py`:

```
sudo bindsnoop.py -t
Tracing binds ... Hit Ctrl-C to end
TIME(s)        PID COMM         PROT ADDR            PORT   OPTS IF
0.000000   2041376 testprocess  TCP  2401:db00:2131: 43173 .....  0
0.000051   2041376 testprocess  TCP  2401:db00:2131: 63381 .....  0
0.000073   2041376 testprocess  TCP  2401:db00:2131: 37789 .....  0
0.000093   2041376 testprocess  TCP  2401:db00:2131: 36263 .....  0
0.000128   2041376 testprocess  TCP  2401:db00:2131: 53143 .....  0
0.000158   2041376 testprocess  TCP  2401:db00:2131: 63743 .....  0
0.296623      2777 testproces2  TCP  2401:db00:2131: 51819 .....  0
0.296669      2777 testproces2  TCP  2401:db00:2131: 44849 .....  0
0.296693      2777 testproces2  TCP  2401:db00:2131: 43833 .....  0
```

Similar question about the commandline flags, looks like you're supporting a subset of bcc-py's, except for the error output flag which was `E` in bcc-py but is now `x`. I'm less concerned about the missing flags than the output differences, but want to be explicit about both choices.

Other than this minor stuff I think this is well-structured and easier to understand than the bcc-py version. Will do another pass tomorrow and test out all the flags on a server, but doubt I'll uncover anything major.

**Comment by chenhengqi** [comment]
> I'm guessing strict compatibility with the `bcc-py` bindsnoop's output format is not a goal? e.g. The example output you provided differs from same invocation of `bindsnoop.py`:
> 
> ```
> sudo bindsnoop.py -t
> Tracing binds ... Hit Ctrl-C to end
> TIME(s)        PID COMM         PROT ADDR            PORT   OPTS IF
> 0.000000   2041376 testprocess  TCP  2401:db00:2131: 43173 .....  0
> 0.000051   2041376 testprocess  TCP  2401:db00:2131: 63381 .....  0
> 0.000073   2041376 testprocess  TCP  2401:db00:2131: 37789 .....  0
> 0.000093   2041376 testprocess  TCP  2401:db00:2131: 36263 .....  0
> 0.000128   2041376 testprocess  TCP  2401:db00:2131: 53143 .....  0
> 0.000158   2041376 testprocess  TCP  2401:db00:2131: 63743 .....  0
> 0.296623      2777 testproces2  TCP  2401:db00:2131: 51819 .....  0
> 0.296669      2777 testproces2  TCP  2401:db00:2131: 44849 .....  0
> 0.296693      2777 testproces2  TCP  2401:db00:2131: 43833 .....  0
> ```
> 
> Similar question about the commandline flags, looks like you're supporting a subset of bcc-py's, except for the error output flag which was `E` in bcc-py but is now `x`. I'm less concerned about the missing flags than the output differences, but want to be explicit about both choices.
> 
> Other than this minor stuff I think this is well-structured and easier to understand than the bcc-py version. Will do another pass tomorrow and test out all the flags on a server, but doubt I'll uncover anything major.

@davemarchevsky  Thanks for the review.

The output format is changed:
1. IMO, the `TIME` column in the original tool does not tell much about the bind event, either wall time or boot time is preferred and useful in this scenario. Also, other libbpf tools use wall time in their output. So I follow this.
2. The `ADDR` column has variable length(v4/v6) so I made it the last column.
3. The `x` cmd flags is already used by other libbpf tools for the same purpose, `execsnoop` for example.
4. For the missing flags, I may do that in a follow up PR.

File: libbpf-tools/bindsnoop.c [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+#include <sys/socket.h>
+#include <time.h>
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "bindsnoop.h"
+#include "bindsnoop.skel.h"
+#include "trace_helpers.h"
+
+#define PERF_BUFFER_PAGES	16
```
[the remark this answers is not in the document]
Comment (chenhengqi) [reply libbpf-tools/bindsnoop.c:23]: OK. 16 pages is adequate for bind events, haven't seen even loss.

**Comment by davemarchevsky** [comment]
[buildbot, test this please]

**Comment by davemarchevsky** [comment]
@chenhengqi Thanks for explaining why the output format differs! I think your reasoning makes sense to me. LGTM.

**Comment by davemarchevsky** [comment]
[buildbot, ok to test]

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Pull request merged**